assertion failed on tpd.TypeApply

## Compiler version

3.0.0-RC3

## Minimized code

```Scala
package example

extension [T](t: T) inline def pi[P <: Tuple](using P): T = ???

inline def env[P <: Tuple, T](op: P ?=> T): P ?=> T = op

@main def Test =
  env { pi[String] }
```

## Output (click arrow to expand)
<details>

```scala
Exception in thread "main" java.lang.AssertionError: assertion failed
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
	at dotty.tools.dotc.ast.tpd$.TypeApply(tpd.scala:50)
	at dotty.tools.dotc.ast.tpd$TreeOps$.appliedToTypeTrees$extension(tpd.scala:949)
	at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:3776)
	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3130)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1377)
	at dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1364)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1377)
	at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1570)
	at dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1642)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:703)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:826)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:727)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:167)
	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:188)
	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:190)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:369)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2611)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2636)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2730)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2822)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2872)
	at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1012)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1016)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2677)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2731)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
	at dotty.tools.dotc.typer.Typer.typedFunctionValue(Typer.scala:1374)
	at dotty.tools.dotc.typer.Typer.makeContextualFunction(Typer.scala:2782)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2728)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$2(ProtoTypes.scala:414)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:370)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:415)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:836)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:836)
	at dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:530)
	at dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:595)
	at dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:433)
	at dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:721)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:835)
	at dotty.tools.dotc.typer.Applications.ApplyTo(Applications.scala:1046)
	at dotty.tools.dotc.typer.Applications.ApplyTo$(Applications.scala:303)
	at dotty.tools.dotc.typer.Typer.ApplyTo(Typer.scala:103)
	at dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:893)
	at dotty.tools.dotc.typer.Applications.realApply$5$$anonfun$4(Applications.scala:972)
	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:2924)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:983)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1021)
	at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:303)
	at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:103)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2669)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2731)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1377)
	at dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1364)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1377)
	at dotty.tools.dotc.typer.Namer.typedAheadRhs$1$$anonfun$1(Namer.scala:1505)
	at dotty.tools.dotc.typer.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:225)
	at dotty.tools.dotc.typer.Namer.typedAheadRhs$2(Namer.scala:1505)
	at dotty.tools.dotc.typer.Namer.rhsType$1(Namer.scala:1513)
	at dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1538)
	at dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1539)
	at dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1550)
	at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1559)
	at dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1642)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:703)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:826)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:727)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:167)
	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:188)
	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:190)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:369)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2611)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2636)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2730)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2822)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2872)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2328)
	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$2(Typer.scala:2657)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2661)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2730)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2822)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2872)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2451)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2702)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2731)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2796)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2800)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2916)
	at dotty.tools.dotc.typer.FrontEnd.liftedTree1$1(FrontEnd.scala:79)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:84)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:43)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:85)
	at dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:120)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:120)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:215)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:223)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:230)
	at dotty.tools.dotc.Run.compileSources(Run.scala:166)
	at dotty.tools.dotc.Run.compile(Run.scala:150)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
	at dotty.tools.dotc.Driver.process(Driver.scala:199)
	at dotty.tools.dotc.Driver.process(Driver.scala:167)
	at dotty.tools.dotc.Driver.process(Driver.scala:179)
	at dotty.tools.dotc.Driver.main(Driver.scala:209)
	at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/ast/tpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/tpd.scala
@@ -41,15 +41,17 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
   def Super(qual: Tree, mixName: TypeName, mixinClass: Symbol = NoSymbol)(using Context): Super =
     Super(qual, if (mixName.isEmpty) untpd.EmptyTypeIdent else untpd.Ident(mixName), mixinClass)
 
-  def Apply(fn: Tree, args: List[Tree])(using Context): Apply = {
-    assert(fn.isInstanceOf[RefTree] || fn.isInstanceOf[GenericApply] || fn.isInstanceOf[Inlined] || fn.isInstanceOf[tasty.TreePickler.Hole])
-    ta.assignType(untpd.Apply(fn, args), fn, args)
-  }
+  def Apply(fn: Tree, args: List[Tree])(using Context): Apply = fn match
+    case Block(Nil, expr) =>
+      Apply(expr, args)
+    case _: RefTree | _: GenericApply | _: Inlined | _: tasty.TreePickler.Hole =>
+      ta.assignType(untpd.Apply(fn, args), fn, args)
 
-  def TypeApply(fn: Tree, args: List[Tree])(using Context): TypeApply = {
-    assert(fn.isInstanceOf[RefTree] || fn.isInstanceOf[GenericApply])
-    ta.assignType(untpd.TypeApply(fn, args), fn, args)
-  }
+  def TypeApply(fn: Tree, args: List[Tree])(using Context): TypeApply = fn match
+    case Block(Nil, expr) =>
+      TypeApply(expr, args)
+    case _: RefTree | _: GenericApply =>
+      ta.assignType(untpd.TypeApply(fn, args), fn, args)
 
   def Literal(const: Constant)(using Context): Literal =
     ta.assignType(untpd.Literal(const))

--- a/tests/neg/i12207.scala
+++ b/tests/neg/i12207.scala
@@ -1,0 +1,8 @@
+package example
+
+extension [T](t: T) inline def pi[P <: Tuple](using P): T = ???
+
+inline def env[P <: Tuple, T](op: P ?=> T): P ?=> T = op
+
+@main def Test =
+  env { pi[String] } // error // error

